### ISSUE TITLE
UID not detected in SAML response during initial login.

### ISSUE THREAD

**Issue by joshdotylmi** [matrix-org/synapse#13092]
### Description

I'm using Azure AD(Microsofts IDP) as a identity provider for Synapse. During the login I'm passing a UID SAML attribute however during the login process this attribute is not registered and I recieved a login error stating that UID was not found during the login process.
The current config I'm using for saml is super basic
```
saml2_config:
  sp_config:
    metadata:
      remote:
        - url: https://login.microsoftonline.com/[my tenant ID/federationmetadata/2007-06/federationmetadata.xml?appid=[ service provider ID]
    service:
      sp:
        allow_unsolicited: true
enable_registration: true
enable_registration_without_verification: true
password_config:
   enabled: false
```

### Steps to reproduce

- Attempted to login to https://riot.im/app with my server
- Clicked "Login with SAML"
- Get redirected to a page stating the uid was not found in the saml response. 

### Homeserver

Personal home server.

### Synapse Version

"server_version":"1.61.0","python_version":"3.9.2"

### Installation Method

Debian packages from packages.matrix.org

### Platform

Virtual Machine running on HyperV
 Debian GNU/Linux 11 (bullseye) x86_64
  Kernel: 5.10.0-13-amd64
Not running in docker container

### Relevant log output

```shell
Attempt #1 Using machine name for the attribute for uid
Formatted Saml response:
<ns0:Assertion xmlns:ns0="urn:oasis:names:tc:SAML:2.0:assertion"
               xmlns:ns1="http://www.w3.org/2000/09/xmldsig#"
               xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               Version="2.0"
               ID="[Removed]"
               IssueInstant="2022-06-16T02:28:50.824Z">
	<ns0:Issuer>[Removed]</ns0:Issuer>
	<ns1:Signature>
		<ns1:SignedInfo>
			<ns1:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<ns1:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
			<ns1:Reference URI="[Removed]">
				<ns1:Transforms>
					<ns1:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<ns1:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</ns1:Transforms>
				<ns1:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
				<ns1:DigestValue>[Removed]</ns1:DigestValue>
			</ns1:Reference>
		</ns1:SignedInfo>
		<ns1:SignatureValue>[Removed]</ns1:SignatureValue>
		<ns1:KeyInfo>
			<ns1:X509Data>
				<ns1:X509Certificate>[Removed]</ns1:X509Certificate>
			</ns1:X509Data>
		</ns1:KeyInfo>
	</ns1:Signature>
	<ns0:Subject>
		<ns0:NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:unspecified">josh.doty@lmi.net</ns0:NameID>
		<ns0:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
			<ns0:SubjectConfirmationData NotOnOrAfter="2022-06-16T03:28:50.636Z"
			                             Recipient="https://[Removed]/_synapse/client/saml2/authn_response"
			                             InResponseTo="id-[Removed]"/>
		</ns0:SubjectConfirmation>
	</ns0:Subject>
	<ns0:Conditions NotBefore="2022-06-16T02:23:50.636Z"
	                NotOnOrAfter="2022-06-16T03:28:50.636Z">
		<ns0:AudienceRestriction>
			<ns0:Audience>https://[Removed]/_synapse/client/saml2/metadata.xml</ns0:Audience>
		</ns0:AudienceRestriction>
	</ns0:Conditions>
	<ns0:AuthnStatement AuthnInstant="2022-06-16T02:28:48.949Z"
	                    SessionIndex="[Removed]">
		<ns0:AuthnContext>
			<ns0:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Unspecified</ns0:AuthnContextClassRef>
		</ns0:AuthnContext>
	</ns0:AuthnStatement>
	<ns0:AttributeStatement>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/tenantid"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">[Removed]</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/objectidentifier"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">[Removed]</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/identityprovider"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="[Removed]</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/claims/authnmethodsreferences"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">http://schemas.microsoft.com/claims/multipleauthn</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/agegroup"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">3</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="urn:oid:0.9.2342.19200300.100.1.1"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">9fasdffdf-2a25-45d2-a170-cf4fc1243d67d29</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="urn:oid:1.2.840.113549.1.9.1.1"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">josh.doty@lmi.net</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="urn:oid:2.16.840.1.113730.3.1.241"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">Josh Doty</ns0:AttributeValue>
		</ns0:Attribute>
	</ns0:AttributeStatement>
</ns0:Assertion>
Error:
2022-06-15 19:28:51,047 - synapse.handlers.saml - 215 - INFO - POST-40 - SAML2 mapped attributes: {}
2022-06-15 19:28:51,047 - synapse.handlers.saml - 424 - WARNING - POST-40 - SAML2 response lacks a 'uid' attestation
2022-06-15 19:28:51,048 - synapse.handlers.saml - 250 - ERROR - POST-40 - Could not map user
Traceback (most recent call last):
  File "/opt/venvs/matrix-synapse/lib/python3.9/site-packages/synapse/handlers/saml.py", line 422, in get_remote_user_id
    return saml_response.ava["uid"][0]
KeyError: 'uid'

During handling of the above exception, another exception occurred:

Traceback (most recent call last):
  File "/opt/venvs/matrix-synapse/lib/python3.9/site-packages/synapse/handlers/saml.py", line 248, in _handle_authn_response
    await self._complete_saml_login(saml2_auth, request, relay_state)
  File "/opt/venvs/matrix-synapse/lib/python3.9/site-packages/synapse/handlers/saml.py", line 275, in _complete_saml_login
    remote_user_id = self._remote_id_from_saml_response(
  File "/opt/venvs/matrix-synapse/lib/python3.9/site-packages/synapse/handlers/saml.py", line 351, in _remote_id_from_saml_response
    remote_user_id = self._user_mapping_provider.get_remote_user_id(
  File "/opt/venvs/matrix-synapse/lib/python3.9/site-packages/synapse/handlers/saml.py", line 425, in get_remote_user_id
    raise MappingException("'uid' not in SAML2 response")
synapse.handlers.sso.MappingException: 'uid' not in SAML2 response


Attempt 2, In this one I changed in Azure AD to pass the uid attribute as simply "uid". Same error in the log as above.
<ns0:Assertion xmlns:ns0="urn:oasis:names:tc:SAML:2.0:assertion"
               xmlns:ns1="http://www.w3.org/2000/09/xmldsig#"
               xmlns:xsi="http://www.w3.org/2001/XMLSchema-instance"
               Version="2.0"
               ID="_5735f290-0380-4ce6-997a-fc0026a79c00"
               IssueInstant="2022-06-16T03:32:44.915Z">
	<ns0:Issuer>[Removed]</ns0:Issuer>
	<ns1:Signature>
		<ns1:SignedInfo>
			<ns1:CanonicalizationMethod Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
			<ns1:SignatureMethod Algorithm="http://www.w3.org/2001/04/xmldsig-more#rsa-sha256"/>
			<ns1:Reference URI="[Removed]">
				<ns1:Transforms>
					<ns1:Transform Algorithm="http://www.w3.org/2000/09/xmldsig#enveloped-signature"/>
					<ns1:Transform Algorithm="http://www.w3.org/2001/10/xml-exc-c14n#"/>
				</ns1:Transforms>
				<ns1:DigestMethod Algorithm="http://www.w3.org/2001/04/xmlenc#sha256"/>
				<ns1:DigestValue>[Removed]</ns1:DigestValue>
			</ns1:Reference>
		</ns1:SignedInfo>
		<ns1:SignatureValue>[Removed]</ns1:SignatureValue>
		<ns1:KeyInfo>
			<ns1:X509Data>
				<ns1:X509Certificate>[Removed]</ns1:X509Certificate>
			</ns1:X509Data>
		</ns1:KeyInfo>
	</ns1:Signature>
	<ns0:Subject>
		<ns0:NameID Format="urn:oasis:names:tc:SAML:1.1:nameid-format:emailAddress">josh.doty@lmi.net</ns0:NameID>
		<ns0:SubjectConfirmation Method="urn:oasis:names:tc:SAML:2.0:cm:bearer">
			<ns0:SubjectConfirmationData NotOnOrAfter="2022-06-16T04:32:44.743Z"
			                             Recipient="https://[Removed]/_synapse/client/saml2/authn_response"
			                             InResponseTo="id-[Removed]"/>
		</ns0:SubjectConfirmation>
	</ns0:Subject>
	<ns0:Conditions NotBefore="2022-06-16T03:27:44.743Z"
	                NotOnOrAfter="2022-06-16T04:32:44.743Z">
		<ns0:AudienceRestriction>
			<ns0:Audience>https://[Removed]/_synapse/client/saml2/metadata.xml</ns0:Audience>
		</ns0:AudienceRestriction>
	</ns0:Conditions>
	<ns0:AuthnStatement AuthnInstant="2022-06-15T21:07:58.227Z"
	                    SessionIndex="[Removed]">
		<ns0:AuthnContext>
			<ns0:AuthnContextClassRef>urn:oasis:names:tc:SAML:2.0:ac:classes:Unspecified</ns0:AuthnContextClassRef>
		</ns0:AuthnContext>
	</ns0:AuthnStatement>
	<ns0:AttributeStatement>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/tenantid"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">[Removed]</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/objectidentifier"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">[Removed]</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/identityprovider"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">[Removed]</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/claims/authnmethodsreferences"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">http://schemas.microsoft.com/ws/2008/06/identity/authenticationmethod/x509</ns0:AttributeValue>
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">http://schemas.microsoft.com/claims/multipleauthn</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="http://schemas.microsoft.com/identity/claims/agegroup"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">3</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="urn:oid:2.16.840.1.113730.3.1.241"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">Josh Doty</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="uid"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">9fasdffdf-2a25-45d2-a170-cf4fc1243d67d29</ns0:AttributeValue>
		</ns0:Attribute>
		<ns0:Attribute Name="Email"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">josh.doty@lmi.net</ns0:AttributeValue>
		</ns0:Attribute>
	</ns0:AttributeStatement>
</ns0:Assertion>


In my attempt 3. Instead of using "uid" used the schema url. Same error


 		<ns0:Attribute Name="http://schemas.xmlsoap.org/ws/2005/05/identity/claims/uid"
		               NameFormat="urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified">
			<ns0:AttributeValue xsi:type="xs:string"
			                    xmlns:xs="http://www.w3.org/2001/XMLSchema">9fasdffdf-2a25-45d2-a170-cf4fc1243d67d29</ns0:AttributeValue>
		</ns0:Attribute>
```

**Comment by anoadragon453** [comment]
Hi @joshdotylmi. It looks like [pysaml2](https://pypi.org/project/pysaml2/) (the saml library we use) is probably looking for something other than just the name of the attribute to extract "uid". As always, digging into this library's code is quite painful - so I don't have an immediate answer for you on what it's looking for.

I have found another issue which seems to offer some advice in this area however. Could you take a look and see if it helps? https://github.com/matrix-org/synapse/issues/5130#issuecomment-514999595

**Comment by joshdotylmi** [comment]
To me the adfs transformation rules seem to be using a prebuilt "uid" format within ADFS rather than something custom. I really wish i could see what the result of the saml responses slipeer was generating with the transformation rules they were using.

Are there any examples of saml responses i could reference @anoadragon453  ? to try and message messages the data azure ad is handing to matrix/synapse.

**Comment by joshdotylmi** [comment]
I mean if i replace the logic check in the saml.py code that's looking for the generated UID with "name" it uses my email in the response.. but then i'm using my full email in the username.
![image](https://user-images.githubusercontent.com/74930105/176963552-00e8979f-bd52-4a75-bc8d-abd3f331509d.png)
![image](https://user-images.githubusercontent.com/74930105/176963571-6b8e6875-30da-45e8-b1bb-23bf6f0578c0.png)

**Comment by richvdh** [comment]
This seems to be something that `pysaml2` massively overcomplicates.

`pysaml2` has a set of "attribute maps" which it uses to map from SAML2 attribute to a Python dict, which it then passes into the application. You can find the maps at https://github.com/IdentityPython/pysaml2/blob/master/src/saml2/attributemaps. Synapse logs the result of that mapping process in the `SAML2 mapped attributes: {...}` line, and the fact that it is empty for you shows that pysaml2 has been unable to map any of the attributes.

`urn:oid:0.9.2342.19200300.100.1.1` *should* map onto `uid` (the mapping is [here](https://github.com/IdentityPython/pysaml2/blob/master/src/saml2/attributemaps/saml_uri.py#L183)), but that's not happening, and I *think* the reason is because the `NameFormat` in your SAML assertion is `urn:oasis:names:tc:SAML:2.0:attrname-format:unspecified` rather than the [expected](https://github.com/IdentityPython/pysaml2/blob/master/src/saml2/attributemaps/saml_uri.py#L53) `urn:oasis:names:tc:SAML:2.0:attrname-format:uri`.

So, there are a few things you can do:
 * Change the attribute names and/or formats presented by AD to match those expected by pysaml2
 * Define your own attribute map via the [`attribute_map_dir`](https://pysaml2.readthedocs.io/en/latest/howto/config.html#attribute-map-dir) setting
 * (probably easiest): set the [`allow_unknown_attributes`](https://pysaml2.readthedocs.io/en/latest/howto/config.html#allow-unknown-attributes) setting under `saml2_config.sp_config.service.sp` which will make pysaml2 pass through anything it doesn't recognise; then either have AD present a `uid` attribute, or change   `saml2_config.user_mapping_provider.config.mxid_source_attribute` to tell synapse to look for something else.

Good luck! Please let us know how you get on - any contribution to the documentation in this area is very welcome.

**Comment by joshdotylmi** [comment]
AH HAH! Thanks @richvdh  That was the key.

I had to put it under saml2_config.sp_config.service
allow_unknown_attributes: true

so it looks like this in the config:

saml2_config:
  sp_config:
    metadata:
      remote:
        - url: https://login.microsoftonline.com/-/federationmetadata/2007-06/federationmetadata.xml?appid=
    allow_unknown_attributes: true

It was only after setting this to true, that my "mxid_source_attribute:" setting was respected. I choose setting it to displayName. which is less than ideal.. but the automatic name generation using full email is too long.

I'm not able to force set my IDP to use urn:oasis:names:tc:SAML:2.0:attrname-format:uri to my knowledge, so i guess trusting unknown attributes is safe.. ish.. 

I'll update the matrix documents with my settings in the future. The saml documentation is a bit sparse right now : p
